### ISSUE TITLE
Investigate removing `AssemblyError` enum

### ISSUE THREAD

**Issue by bobbinth** [0xMiden/miden-vm#1431]
Currently, public API of the assembler returns `Report` errors. However, internally in the assembler we generate `AssemblyError` in quite a few places. We should investigate whether we can get rid of `AssemblyError` entirely and just use `Report `everywhere instead.

**Comment by bitwalker** [comment]
Definitely - this is an easy one to do, but as a pre-requisite I'd like to add a macro for creating ad-hoc `Report`s (basically `miette::miette!`, but our own, so that we don't have to export `miette` from our public API). After that, all of the existing places where we return `AssemblyError` can be converted to use that.

We might still want a concrete error type that implements `Diagnostic`, for errors that are used in multiple places, so that we don't have to duplicate the diagnostic information multiple times, but I think there are still several `AssemblyError` variants which are used in basically one place, but I'd have to double check.

**Comment by bobbinth** [comment]
> This is an example of an error where using `Report` and ad-hoc diagnostics works better than a concrete error type like `AssemblyError`. This error sets the stage for what is wrong, but is sort of incomplete from the user perspective. Ideally, you'd have the primary label span the procedure body, with a secondary label spanning the first decorator, with a message like `'<opcode>' is a decorator, which cannot appear in a procedure body without at least one non-decorator instruction`, with help text attached to the diagnostic like `try adding a 'nop' instruction at the end of the procedure body to anchor the decorators`.
>
> By using `AssemblyError` for things like this, we make it difficult (if not impossible) to construct these kind of helpful diagnostics, which is why most APIs are designed to return `Report` rather than `AssemblyError`.
>
> Your call on whether you want to make the change, but we should try to get in the habit of thinking about how these errors will be seen and acted on by users, particularly those that don't have as strong a familiarity with Miden Assembly, and may not understand the difference between things like decorators and instructions, since they appear the same way in the source code.

_Originally posted by @bitwalker in https://github.com/0xPolygonMiden/miden-vm/pull/1466#discussion_r1755153155_

**Comment by plafer** [comment]
See https://github.com/0xPolygonMiden/miden-vm/pull/1588#discussion_r1867423753 for potential mismatch between using `thiserror`'s `#[source]` and miette's `Report`. Just something to be aware of when addressing this issue.

**Comment by Fumuran** [comment]
>as a pre-requisite I'd like to add a macro for creating ad-hoc Reports (basically miette::miette!, but our own, so that we don't have to export miette from our public API

@bitwalker I understand the idea of doing so, but how do you think it should be implemented? I mean, if I create a thin wrapper around `miette::miette!` macro, a user will have to import `miette!` into the file where it will be used anyway. Going deeper (and copying the implementation of `miette!` macro) will require to import internal `miette` structs, which sounds even worse. Should we essentially copy the entire implementation of the `miette!` just to remove it from our exports? 

>We might still want a concrete error type that implements Diagnostic, for errors that are used in multiple places

I replaced variants which were used only once with ad-hoc `Report` using (for now) `miette!` macro, but more than a half of the `AssemblyError` variants are used in multiple places. In that case, should we leave the `AssemblyError` enum with this variants to avoid duplication?

**Comment by plafer** [comment]
Coming back to this, I'm not sure why we want to remove `AssemblyError` - pretty much for the reasons @Fumuran talked about above. I see the main benefit of `AssemblyError` to be the ability to reuse errors - this is effectively what I did in #1717 with `MemoryError`. And if anything, I find it helpful to have `AssemblyError` a an amalgamation of all errors returned, with a description of how they are being output as a `Diagnostic`.

Is this issue still relevant?

**Comment by bobbinth** [comment]
@bitwalker - you had some pretty strong reasons for this. Should we pick this up again?

**Comment by bitwalker** [comment]
@bobbinth I think in general we've moved to more targeted usage of `AssemblyError`, and broader usage of `Report`, which is the direction I wanted to head in.

I still think `AssemblyError`, in its current state, is bloated and can be drastically simplified, let me demonstrate:

* `Empty` - used in a single place, when recomputing the module graph. IMO, we probably should have an error type just for the things that can go wrong in the module graph, i.e. `ModuleGraphError`. That alone would simplify `AssemblyError` a bunch, and would make it clearer where those errors come from.
* `Failed` - this exists to allow us to generate diagnostics which want to emit labels in different source files. It is also used only within the module graph currently. This is functionality I'd like to merge into `Report` anyway, but I haven't had time to get around to it. For now, I'd suggest we move this to `ModuleGraphError`
* `Cycle` - used in two places, both in the call graph data structure that is used by the module graph, so this too should move to `ModuleGraphError`
* `ConflictingDefinitions` - used in one place, in the `MastForestBuilder`, no reason this can't be an ad-hoc `Report`
* `DuplicateModule` - used in one place, in the module graph, no reason this can't be an ad-hoc `Report`
* `UndefinedModule` - used twice, in the module graph, this too could be an ad-hoc `Report`, but moving it to `ModuleGraphError` would be fine as well
* `InconsistentNamespace` - not even used
* `InvalidSysCallTarget` - the first variant that is used in multiple places, including the assembler itself. Makes sense to keep this one.
* `LocalWordIndex` - used in two places to catch invalid procedure local indices. However, this is already validated in the parser/semantic analysis (and if it isn't, it should be), so I don't believe this needs to even be handled in the assembler (except as an assertion, if anything).
* `InvalidNumProcLocals` - used in one place, no reason this couldn't be an ad-hoc `Report`. This too is already validated by the parser/semantic analysis IIRC, but regardless, this variant isn't needed.
* `InvalidU8Param` - used in the lowering of multiple instructions, but is also something that is validated by the parser/semantic analysis (or should be), so I would expect these to just be assertions by the time we hit assembly, but we do get some benefit by rendering the associated source code by using error diagnostics here, so I think this is fine to keep.
* `CallerOutsideOfKernel` - used in one place, no reason this couldn't be an ad-hoc `Report`
* `EmptyProcedureBodyWithDecorators` - used in one place, no reason this couldn't be an ad-hoc report
* `ProcLocalsNotDeclared` - used in one place, no reason this couldn't be an ad-hoc report
* `Other` - exists to wrap `Report`s in a `AssemblyError`, since `Report` itself cannot implement the `Diagnostic` trait due to trait coherence rules
* `Forest` - exists to wrap `MastForestError`, so we'd continue to need this while `AssemblyError` exists

**Comment by bobbinth** [comment]
@Fumuran - could you take a stab at simplifying the `AssemblyError` as @bitwalker outlined above.

**Comment by bitwalker** [comment]
@bobbinth @Fumuran I can actually tackle this as a follow-on to my other `ModuleGraph` changes, since I'll be in there anyway

**Comment by bobbinth** [comment]
> [@bobbinth](https://github.com/bobbinth) [@Fumuran](https://github.com/Fumuran) I can actually tackle this as a follow-on to my other `ModuleGraph` changes, since I'll be in there anyway

Sounds good! I'll re-assign this to you then. Thank you!

**Comment by Fumuran** [comment]
@bitwalker I could push the refactoring I already did, but probably it's too old to be valuable. But let me know if it helps.

**Comment by bitwalker** [comment]
I'll keep that in mind @Fumuran, thanks! I'm guessing with the changes I have in mind, it's about equal work to adapt your branch as making them against `next`, but if you think there are some things in there you'd like to see adopted, let me know!

**Comment by bobbinth** [comment]
Closed by #1881.
